### PR TITLE
Added Group migration functionality to Baton plugin

### PR DESCRIPTION
**Pull request by jacksondelametter**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -56,18 +56,22 @@ The following example configures the migration to only look in the `./src/main/r
...
-    "implementation": "org.technologybrewery.baton.example.FooToBarMigration",
-    "fileSets": [
-      {
-        "includes": ["**/legacy/*.foo"],
-        "excludes": ["original-specification-example.foo"]
-      }
-    ]
-  }
+    {
+        "group": "group_1",
```
Comment (d-ryan-ashcraft) [inline README.md]: A: very minor, but recommend we make this `"group-1"` (dash vs. underscore for consistency with maven norms) or something more functionally meaningful (e..g, `"upgrade-abc-migrations"`)

File: baton-maven-plugin/src/main/java/org/technologybrewery/baton/BatonMojo.java [hunk 2]
```diff
@@ -230,37 +232,17 @@ protected void loadMigrations() {
                         migrationsStream,
                         tempMigrationsFile.toPath(),
                         StandardCopyOption.REPLACE_EXISTING);
```
Comment (d-ryan-ashcraft) [inline baton-maven-plugin/src/main/java/org/technologybrewery/baton/BatonMojo.java:234]: A: not a big deal, but I think inlining this change creates more cognitive friction to someone reading the code.  It was super clear what was happening before, b/c of the method name.  Now you have to read two lines and interpret on the fly to understand.  Sometimes it is not how many lines, but how easy is it to follow.

A: line 236 should be indented, if this code stays this way.

File: baton-maven-plugin/src/main/java/org/technologybrewery/baton/GroupSummary.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package org.technologybrewery.baton;
+
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (d-ryan-ashcraft) [inline baton-maven-plugin/src/main/java/org/technologybrewery/baton/GroupSummary.java:5]: S: I know it seems obvious, but sometimes writing the comment identifies how names may or may not make sense.  Our goal is for all non getter/setter public classes and methods to have a comment.

File: baton-maven-plugin/src/main/java/org/technologybrewery/baton/config/GroupTarget.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.technologybrewery.commons.json.AbstractValidatedElement;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
```
Comment (d-ryan-ashcraft) [inline baton-maven-plugin/src/main/java/org/technologybrewery/baton/config/GroupTarget.java:11]: S: comment please

File: baton-maven-plugin/src/test/resources/specifications/baton-maven-plugin.feature [hunk 5]
```diff
@@ -1,12 +1,64 @@
+@baton
 Feature: Validate the definition of a migration target
 
-  Scenario: A simple migration can be defined and executed
-    Given a migration target with a name and a implementation
+  Scenario: Group migrations can be defined and executed in their order
+    Given a migration file 
+        """
+        [
```
Comment (d-ryan-ashcraft) [inline baton-maven-plugin/src/test/resources/specifications/baton-maven-plugin.feature]: S: normally, we are happy to hide some of the detail here.  
```
Given a migration files with <important part for this scenario>" (dirty work or mapping to a file in the step)
...

// So here, maybe something like:
Given a mogration with groups in the order of A, B, C
```

For this test, the only thing the reader cares about is the order of the groups.  But they have to read through the rest of it before realizing that.  We want to be focused instead.

File: baton-maven-plugin/src/test/resources/specifications/migration-json.feature [hunk 6]
```diff
@@ -2,38 +2,41 @@
...
-      |                          |              | **/*          |                |
-      |                          |              | *.xml, *.java |                |
-      |                          |              |               | true           |
-      |                          |              |               | false          |
-      |                          |              |               |                |
\ No newline at end of file
+    | group    | directory                | includes     | excludes      | followSymLinks |
+    | Foo      | ./src/my-random-location |              |               |                |
+    | Bar      | ./src/main/resources     |              |               |                |
+    | Foo_Bar  |                          | **/*.foo     |               |                |
```
Comment (d-ryan-ashcraft) [inline baton-maven-plugin/src/test/resources/specifications/migration-json.feature]: A: same comment about underscores vs. dashes.  Maven typically prefers dashes - it's really easy to mix them, so staying super consistent will help mitigate that.

File: examples/example-migration-configuration/src/main/resources/migrations.json [hunk 7]
```diff
@@ -1,20 +1,44 @@
 [
   {
-    "name": "test-migration-a",
-    "implementation": "org.technologybrewery.baton.LoggingSuccessfulMigration",
-    "fileSets": [
-      {
-        "includes": ["**/*.txt"]
-      }
+    "group": "group-1",
```
Comment (d-ryan-ashcraft) [inline examples/example-migration-configuration/src/main/resources/migrations.json:3]: A: case in point...

**Review by d-ryan-ashcraft — COMMENTED** [review]
Looks good - some small items inlined.

File: README.md (hunk 1 above)
Comment (jacksondelametter) [reply README.md]: Fixed.

File: baton-maven-plugin/src/main/java/org/technologybrewery/baton/BatonMojo.java (hunk 2 above)
Comment (jacksondelametter) [reply baton-maven-plugin/src/main/java/org/technologybrewery/baton/BatonMojo.java:234]: Fixed.

File: baton-maven-plugin/src/main/java/org/technologybrewery/baton/GroupSummary.java (hunk 3 above)
Comment (jacksondelametter) [reply baton-maven-plugin/src/main/java/org/technologybrewery/baton/GroupSummary.java:5]: Added comments.

File: baton-maven-plugin/src/main/java/org/technologybrewery/baton/config/GroupTarget.java (hunk 4 above)
Comment (jacksondelametter) [reply baton-maven-plugin/src/main/java/org/technologybrewery/baton/config/GroupTarget.java:11]: Added comment

File: baton-maven-plugin/src/test/resources/specifications/migration-json.feature (hunk 6 above)
Comment (jacksondelametter) [reply baton-maven-plugin/src/test/resources/specifications/migration-json.feature]: Fixed.

File: baton-maven-plugin/src/test/resources/specifications/baton-maven-plugin.feature (hunk 5 above)
Comment (jacksondelametter) [reply baton-maven-plugin/src/test/resources/specifications/baton-maven-plugin.feature]: Test has been modified to focus on the group ordering.

**Review by d-ryan-ashcraft — APPROVED** [review]

**Pull request merged**